src/simplewallet/simplewallet.cpp

```diff
@@ -1090,6 +1074,71 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
       bool r = new_wallet(vm, m_recovery_key, m_restore_deterministic_wallet, m_non_deterministic, old_language);
       CHECK_AND_ASSERT_MES(r, false, tr("account creation failed"));
     }
+    if (!m_restore_height && m_restoring)
+    {
+      uint32_t version;
+      if (!try_connect_to_daemon(false, &version))
+      {
+        fail_msg_writer() << tr("failed to connect to the daemon");
+        return false;
+      }
+      while (true)
+      {
+        std::string heightstr = command_line::input_line(
+          version < 0x1006 ? "Restore from specific blockchain height (optional, default 0): ":
+                             "Restore from specific blockchain height (optional, default 0),\nor alternatively from specific date (YYYY-MM-DD): ");
+        if (std::cin.eof())
+          return false;
+        if (heightstr.empty())
+        {
+          m_restore_height = 0;
+          break;
+        }
+        try
+        {
+          m_restore_height = boost::lexical_cast<uint64_t>(heightstr);
+          break;
+        }
+        catch (boost::bad_lexical_cast &)
+        {
+          if (version < 0x1006)
+          {
+            fail_msg_writer() << tr("bad m_restore_height parameter: ") << heightstr;
+            continue;
+          }
+          if (heightstr.size() != 10 || heightstr[4] != '-' || heightstr[7] != '-')
+          {
+            fail_msg_writer() << tr("date format must be YYYY-MM-DD");
+            continue;
+          }
+          uint16_t year;
+          uint8_t month;  // 1, 2, ..., 12
+          uint8_t day;    // 1, 2, ..., 31
+          try
+          {
+            year  = boost::lexical_cast<uint16_t>(heightstr.substr(0,4));
+            month = boost::lexical_cast<uint16_t>(heightstr.substr(5,2));
+            day   = boost::lexical_cast<uint16_t>(heightstr.substr(8,2));
```

Comment:
That's fine. For scanf, you'd use, eg:
if (sscanf(heightstr.c_str(), "%04u-%02u-%02u", &year, &month, &day) < 3) { error }
That seems a lot simple to me than a dozen lines, but your code seems alright otherwise, so it's fine. No need for static casts.